### PR TITLE
Add user id to marker entity

### PR DESCRIPTION
**Pull request by NoyFlat**
The idToken of the user is sent to the post request
If the idToken is an empty string, there is no user
Added verification of the token in marker servlet
Added the userId field to the marker as optional
Added the userId property to marker entity if user uploaded the marker

### CODE REVIEW

File: capstone-project/src/main/angular-webapp/src/app/info-window/info-window.component.ts [hunk 1]
```diff
@@ -31,7 +31,8 @@ export class InfoWindowComponent implements OnInit {
     this.animal = animalValue;
     this.description = descriptionValue;
     this.reporter = reporterValue;
-    this.submitEvent.emit({animal: animalValue, description: descriptionValue, reporter: reporterValue})
+    let idToken = (this.user ? this.user.idToken : "")
```
Comment (NaamaHolzer) [inline capstone-project/src/main/angular-webapp/src/app/info-window/info-window.component.ts:34]: Can this be const?

**Review by NaamaHolzer — APPROVED** [review]

File: capstone-project/src/main/angular-webapp/src/app/info-window/info-window.component.ts [hunk 2]
```diff
@@ -31,7 +31,8 @@ export class InfoWindowComponent implements OnInit {
     this.animal = animalValue;
     this.description = descriptionValue;
     this.reporter = reporterValue;
-    this.submitEvent.emit({animal: animalValue, description: descriptionValue, reporter: reporterValue})
+    const idToken = (this.user ? this.user.idToken : "")
```
Comment (orarch) [inline capstone-project/src/main/angular-webapp/src/app/info-window/info-window.component.ts:34]: The parentheses are redundant. 

You can use optional chaining operator: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: capstone-project/src/main/angular-webapp/src/app/info-window/info-window.component.ts [hunk 3]
```diff
@@ -31,7 +31,8 @@ export class InfoWindowComponent implements OnInit {
...
+    this.submitEvent.emit({animal: animalValue, description: descriptionValue, reporter: reporterValue, userToken: idToken})
```
Comment (orarch) [inline capstone-project/src/main/angular-webapp/src/app/info-window/info-window.component.ts:35]: I don't think it makes sense for the info window component to pass the user ID to the map component.

That's why you have the UserService, just inject it in the map component

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 4]
```diff
@@ -115,6 +153,12 @@ private static void updateMarker(Marker marker) throws EntityNotFoundException {
         // Find the entity of the marker that needs to be updated using its ID
         Key markerEntityKey = KeyFactory.createKey("Marker", marker.getId());
         Entity markerEntity = datastore.get(markerEntityKey);
+        // If markerEntity has userId, the updated marker should have the same userId
+        Optional<String> userId = Optional.empty();
+        if (markerEntity.hasProperty("userId")){
+            userId = Optional.of((String) markerEntity.getProperty("userId"));
+        }
+        marker.setUserId(userId);
```
Comment (orarch) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:161]: When can that happen?

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 5]
```diff
@@ -86,6 +99,31 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         }
     }
 
+    /** Verifies the idToken and returns the user ID if token is verified */
+    private static Optional<String> verifyToken(String idTokenString) throws IOException {
+        // If there's no user, the idTokenString received is an empty string
+        if (idTokenString.equals("")){
+            return Optional.empty();
+        }
+		try {
```
Comment (orarch) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:108]: fix indentations

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java (hunk 4 above)
Comment (NoyFlat) [reply capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:161]: Right now, user doesn't have to sign in to upload a marker

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java (hunk 5 above)
Comment (NoyFlat) [reply capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:108]: In the editor it looks fine, I don't know why it looks weird here 
image

File: capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 6]
```diff
@@ -86,6 +99,31 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+                .build();
+            GoogleIdToken idToken;
+            idToken = verifier.verify(idTokenString);
+            if (idToken != null) {
+                Payload payload = idToken.getPayload();
+                return Optional.of(payload.getSubject());
+            }
+        } catch (GeneralSecurityException e) {
+            System.out.println("idToken unauthorized");
+            e.printStackTrace();
```
Comment (orarch) [inline capstone-project/src/main/java/com/google/sps/servlets/MarkerServlet.java:122]: https://www.tutorialspoint.com/servlets/servlets-debugging.htm

**Review by orarch — APPROVED** [review]

**Pull request merged**